pkg/river/encoding/block.go

```diff
@@ -0,0 +1,86 @@
+package encoding
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/grafana/agent/pkg/river/internal/rivertags"
+	"github.com/grafana/agent/pkg/river/internal/value"
+)
+
+// BlockField represents the json representation of a river block.
+type BlockField struct {
+	Field           `json:",omitempty"`
+	ID              string        `json:"id,omitempty"`
+	Label           string        `json:"label,omitempty"`
+	Body            []interface{} `json:"body,omitempty"`
+	attributeFields []*AttributeField
+	blockFields     []*BlockField
+}
+
+func newBlock(val value.Value, f rivertags.Field) (*BlockField, error) {
+	bf := &BlockField{}
+	return bf, bf.convertBlock(val, f)
+}
+
+// MarshalJSON implements json marshaller.
+func (bf *BlockField) MarshalJSON() ([]byte, error) {
+	bf.Body = make([]interface{}, 0)
+	for _, x := range bf.attributeFields {
+		bf.Body = append(bf.Body, x)
+	}
+	for _, x := range bf.blockFields {
+		bf.Body = append(bf.Body, x)
+	}
+	type temp BlockField
+	return json.Marshal((*temp)(bf))
+}
+
+func (bf *BlockField) isValid() bool {
+	if bf == nil {
+		return false
+	}
+	return len(bf.blockFields)+len(bf.attributeFields) > 0
+}
+
+func (bf *BlockField) convertBlock(val value.Value, f rivertags.Field) error {
+	if val.Reflect().Kind() != reflect.Struct {
+		return fmt.Errorf("convertBlock cannot work on interface or slices")
+	}
+	if val.Interface() == nil {
+		return nil
+	}
+	if val.Reflect().IsZero() {
```

Comment:
Can you bring out `val.Reflect()` into a variable? It's called a few times in this method and I'm worried about the cost of copying the value each time.